File: SourceCppProject/sources/Services/AppIntelligentWorker.cpp
#include "AppIntelligentWorker.hpp"

#include "../Models/SourcesExplorer.hpp"
#include "../Printers/ConsoleOutputPrinter.hpp"
#include "../Printers/FileOutputPrinter.hpp"
#include "../Printers/WebOutputPrinter.hpp"
#include "AppDatetimeService.hpp"


void AppIntelligentWorker::performTask(const AppLaunchConfiguration& configuration)
{
    const AppDatetimeService datetime = AppDatetimeService();

    const auto arguments = configuration.getArgumentsOfSources();
    const SourcesExplorer explorer(arguments);

    if (isNeededIncreasePerformance(arguments.size()))
    {
        const auto artifacts = explorer.getArtifactsFromSourcesAsMapBasedContent();

        showResults(artifacts);
        saveResultsToTextFile(artifacts);
        saveResultsToWebPage(artifacts);
    }
    else
    {
        const auto artifacts = explorer.getArtifactsFromSourcesAsSetBasedContent();

        showResults(artifacts);
        saveResultsToTextFile(artifacts);
        saveResultsToWebPage(artifacts);
    }

    datetime.printExecutionElapsedTime();
}

bool AppIntelligentWorker::isNeededIncreasePerformance(const unsigned int count_of_sources)
{
    return count_of_sources > m_CriticalPerformanceThreshold;
}

void AppIntelligentWorker::showResults(const LinesResultMapBasedView& result_view)
{
    ConsoleOutputPrinter().printArtifacts(result_view);
}

void AppIntelligentWorker::showResults(const LinesResultSetBasedView& result_view)
{
    ConsoleOutputPrinter().printArtifacts(result_view);
}

void AppIntelligentWorker::saveResultsToTextFile(const LinesResultMapBasedView& result_view)
{
    FileOutputPrinter result_file;
    result_file.printArtifacts(result_view);
}

void AppIntelligentWorker::saveResultsToTextFile(const LinesResultSetBasedView& result_view)
{
    FileOutputPrinter result_file;
    result_file.printArtifacts(result_view);
}

void AppIntelligentWorker::saveResultsToWebPage(const LinesResultMapBasedView& result_view)
{
    WebOutputPrinter result_page;
    result_page.printArtifacts(result_view);
}

void AppIntelligentWorker::saveResultsToWebPage(const LinesResultSetBasedView& result_view)
{
    WebOutputPrinter result_page;
    result_page.printArtifacts(result_view);
}

File: TextLinesComparing.Library/AppIntelligentWorker.cs
namespace TextLinesComparing.Library;

public static class AppIntelligentWorker
{
    const int CriticalPerformanceThreshold = 2;

    public static void PerformTask(string[] args)
    {
        AppDatetimeService datetime = new();

        List<string> arguments = AppLaunchConfiguration.GetArgumentsOfSources(args);
        SourcesExplorer explorer = new(arguments);

        if (IsNeededIncreasePerformance(arguments.Count))
        {
            LinesResultView<LinesStorageMap> artifacts = 
                explorer.GetArtifactsFromSourcesAsMapBasedContent();

            ShowResults(artifacts);
            SaveResultsToTextFile(artifacts);
            SaveResultsToWebPage(artifacts);
        }
        else
        {
            LinesResultView<LinesStorageSet> artifacts = 
                explorer.GetArtifactsFromSourcesAsSetBasedContent();

            ShowResults(artifacts);
            SaveResultsToTextFile(artifacts);
            SaveResultsToWebPage(artifacts);
        }

        datetime.PrintExecutionElapsedTimeMessage();
    }



    private static bool IsNeededIncreasePerformance(int count_of_sources) => 
        count_of_sources > CriticalPerformanceThreshold;

    private static void ShowResults(LinesResultView<LinesStorageMap> result_view)
    {
        new OutputConsolePrinter().PrintArtifacts(result_view);
    }

    private static void ShowResults(LinesResultView<LinesStorageSet> result_view)
    {
        new OutputConsolePrinter().PrintArtifacts(result_view);
    }

    private static void SaveResultsToTextFile(LinesResultView<LinesStorageMap> result_view)
    {
        new OutputFilePrinter().PrintArtifacts(result_view);
    }

    private static void SaveResultsToTextFile(LinesResultView<LinesStorageSet> result_view)
    {
        new OutputFilePrinter().PrintArtifacts(result_view);
    }

    private static void SaveResultsToWebPage(LinesResultView<LinesStorageMap> result_view)
    {
        new OutputWebPrinter().PrintArtifacts(result_view);
    }

    private static void SaveResultsToWebPage(LinesResultView<LinesStorageSet> result_view)
    {
        new OutputWebPrinter().PrintArtifacts(result_view);
    }
}
